scipy/stats/_binomtest.py

```diff
@@ -327,7 +328,19 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
     return result
 
 
-def binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
+def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
```

Comment:
@mdhaber - any update on this PR? I think I addressed all comments apart from the one about using an existing bisection function. I still don't see a good function that solves all requirements. My concern with using scipy.optimize.bisect is that it is meant for continuous functions. The binom.pmf function we need to do the binary search on is discrete, returning 0 for non-integer values. We could cast the values to integers, but this will make it a step function. Now, applying scipy.optimize.bisect on a step function can return any value in the range of the step. For ex: if fn(2)=0.1 and fn(3)=0.2 and we're searching for 0.15, it could return anything between 2 and 3 (in fact, in one such test it returns 3.000000000000682; what if tomorrow it starts returning 2.9999999958?). There is no way to always guarantee that it'll return either 2 or 3 (now and in all future versions of the function). And a change in behavior could break this function. Please advise - you still want to use it here?

Reply:
I thought that might be the trouble but i wanted to play with it to understand better. I haven't had time yet.
Without fully understanding, does that mean it c will get you within 1 of the correct integer? Is it possible to check the adjacent integer after it returns to see if that is the correct one?

Reply:
I played around with this. Something like this:
```python3
def _binary_search_for_binom_tst2(d, lo, hi):
    def f(i, d):
        i_low = np.minimum(np.floor(i), n-1)
        a, b = binom.pmf((i_low, i_low+1), n, p)
        c = a + (i - i_low) * (b - a)
        return c - d
    return root_scalar(f, bracket=(lo, hi), args=(d,)).root
```
would work. It just solves for the float `k` at which a linearly-interpolated binomial PMF would equal `d`, then you `np.floor` or `np.ceil` the result depending on the side you're looking at. Otherwise it works similarly to what you have, although I might have missed an edge case. It is usually a bit slower, but it also scales well.

I'm not sure which I prefer. If the one above were _faster_, I'd prefer it, but I suppose the default is to leave what you have alone. I think @WarrenWeckesser should take a look at this PR when he can, though, because he wrote `binomialtest`.

Reply:
Actually, we can stop when `d` is between `a` and `b`:
```python3
def _binary_search_for_binom_tst2(d, lo, hi):
    def f(k, d):
        k_low = np.minimum(np.floor(k), n-1)
        a, b = binom.pmf((k_low, k_low+1), n, p)
        return (0 if (a <= d <= b) or (a >= d >= b)
                else a + (k - k_low) * (b - a) - d)  # interpolate
    res = root_scalar(f, bracket=(lo, hi), args=(d,))
    return res.root
```
With that change, it begins to compete with `_binary_search_for_binom_tst`.

Reply:
I was going to try this over the next few days, but I'm glad you did :)
I tried your function and am having trouble:
>n=10; p=0.4
> d=0.22 #Its between pmf(3) and pmf(4)
>_binary_search_for_binom_tst2(d,0,5)

Return ValuError: f(a) and f(b) must have different signs (the original binary search version returns 3 with asc_order=True).

I expect this to be a bit slower since the root_scalar function doesn't know that the function is monotonically decreasing or monotonically increasing.

Reply:
I think it works well for the right side of the distribution but sometimes not for the left side. I'll try to fix it as well. Another fear I have (if I understand the function correctly) is that you set the function to 0 over the entire valid range that contains d (right?). In most cases, the root_scalar should return something in between that range. Then ceiling or floor will work. But, it could one day return the end points as well. For example, if pmf(3)<d<pmf(4); it could return anything between 3.0 and 4.0, inclusive (correct?).. then if you do ceil(3.2) you get 4 but if you do ceil(3.0), you get 3.0. This could be a rare edge case that never manifests though.

Reply:
I suppose we need to think about what we _want_ to happen when d equals one of the PMFs exactly (is it possible for values to the left and right of the mode to be equal?) and choose whether the inequality should be strict or not based on that.

Reply:
Do you know what the answer is? (Which index do we want to return when the PMF is equal on both sides?)

Reply:
If you look at the existing code, in the case of two-sided, they first calculate d, the pmf at input k value:

>d = binom.pmf(k,np)

Then they multiply d by some relative error:
> rerr = 1 + 1e-7

Then they look for the number of terms on the other side where the PMF <= d*rerr

> y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)

Not sure why they do this but with multiplying the rerr, but it effectively makes it a < condition instead of <=.

In real life, it'll have to be a perfect storm where d*rerr will happen to equal exactly a PMF on the other side (if it were just d without the multiplication with rerr, just a binomial PMF with p=0.5 would exercise this condition).

What I can say is that my code preserves existing behavior.

> n=10; p=0.5
> i=np.arange(np.ceil(p * n),n+1)
> y = np.sum(binom.pmf(i,n,p)<=binom.pmf(7,n,p),axis=0)

This returns 3. And my code:

> _binary_search_for_binom_tst(lambda x1:binom.pmf(x1,n,p),binom.pmf(7,n,p),0,np.floor(p*n),True)+1

Also returns 3.

Reply:
While @mdhaber's approach reuses an existing root finder, I think the implementation of `_binary_search_for_binom_tst` is simple enough that I'm OK with using it as it is.  It stays in the domain of integer indices, whereas using `root_scalar` requires a conversion to a continuous dependent variable, so it feels a bit more complicated than necessary.   Sure, it is actually pretty simple, and if using `root_scalar` turned out to be significantly faster, I'd say go with it.  But if it is borderline, then I'm fine with the existing bisection code.  Perhaps we can defer further investigation of optimizing the performance to a follow-up PR.

Reply:
Sounds good. I can't verify that the `_binary_search_for_binom_tst` code will do the right thing in all cases, but if someone else can, the rest of the PR looks good to me.

Reply:
@mdhaber - I think the tests I've added are pretty extensive and expose it to almost all conditions conceivable and ensure it maintains existing behavior (with the added benefits of speed and memory). I covered d being on the high side and search on low, d being on low, search on high, d being equal to input on the other side (added explicit test on binary search for this since the condition is very difficult to exercise in usage of binomtest) and monotonic PMF. I can't think of anything else but if someone else can, happy to add more tests to verify.

Reply:
Oops, I agree that the tests are very good. I believe it works very well, and I certainly can't think of counter-examples. I just meant to encourage someone else to take a close look.